add channels last support for ChannelShuffle

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70897
* #70665
* #68101
* #58348
* #55864
* #55584
* #69067
* #50573
* __->__ #50247

Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

mingfeima committed:
add channels last support for ChannelShuffle

[ghstack-poisoned]

mingfeima commented:
This patch adds native support of channels last memory format for `nn.ChannelShuffle`.
Currently `nn.ChannelShuffle` cannot propagate input memory format, channels last would be treated as non-contiguous NCHW tensor.

From performance perspective, ChannelShuffle favors NCHW over NHWC, since on NHWC it would end up with matrix transpose. I made fast paths with vectorized transpose for groups = 2, 4, 8, 16, ... so that NHWC has similar perf as NCHW.

### Performance Benchmarking
I tested pytorch operator benchmark for [ChannelShuffle](https://github.com/pytorch/pytorch/blob/master/benchmarks/operator_benchmark/pt/channel_shuffle_test.py), Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz, single core.

Unit: (us)
Jemalloc is applied, the benchmark has quite a large memory footprint.

The benchmark naming is self explanatory.

#### short tag:
```bash
python -m pt.channel_shuffle_test --omp_num_threads 1 --mkl_num_threads 1
```


Name: (tag short) | before | after | speedup
-- | -- | -- | --
channel_shuffle_batch_size2_channels_per_group16_height16_width16_groups2_channel_lastTrue | 15.597 | 5.965 | 2.61
channel_shuffle_batch_size2_channels_per_group16_height16_width16_groups2_channel_lastFalse | 8.208 | 4.433 | 1.85
channel_shuffle_batch_size2_channels_per_group32_height32_width32_groups2_channel_lastTrue | 102.936 | 31.916 | 3.23
channel_shuffle_batch_size2_channels_per_group32_height32_width32_groups2_channel_lastFalse | 37.120 | 28.515 | 1.30
channel_shuffle_batch_size4_channels_per_group32_height32_width32_groups4_channel_lastTrue | 427.871 | 160.150 | 2.67
channel_shuffle_batch_size4_channels_per_group32_height32_width32_groups4_channel_lastFalse | 168.129 | 159.821 | 1.05
channel_shuffle_batch_size4_channels_per_group64_height64_width64_groups4_channel_lastTrue | 13326.980 | 1902.588 | 7.00
channel_shuffle_batch_size4_channels_per_group64_height64_width64_groups4_channel_lastFalse | 1977.063 | 1943.890 | 1.02
channel_shuffle_batch_size8_channels_per_group64_height64_width64_groups8_channel_lastTrue | 67471.954 | 11689.359 | 5.77
channel_shuffle_batch_size8_channels_per_group64_height64_width64_groups8_channel_lastFalse | 9851.476 | 9995.010 | 0.99
channel_shuffle_batch_size16_channels_per_group64_height64_width64_groups16_channel_lastTrue | 299455.756 | 52310.426 | 5.72
channel_shuffle_batch_size16_channels_per_group64_height64_width64_groups16_channel_lastFalse | 42148.276 | 42693.034 | 0.99


#### long tag:
```bash
python -m pt.channel_shuffle_test --omp_num_threads 1 --mkl_num_threads 1 --tag_filter long
```


Name: (tag long) | before | after | speedup
-- | -- | -- | --
channel_shuffle_batch_size4_channels_per_group32_height32_width32_groups4_channel_lastTrue | 456.409 | 162.546 | 2.81
channel_shuffle_batch_size4_channels_per_group32_height32_width32_groups4_channel_lastFalse | 172.105 | 162.034 | 1.06
channel_shuffle_batch_size4_channels_per_group32_height32_width32_groups8_channel_lastTrue | 1433.409 | 328.539 | 4.36
channel_shuffle_batch_size4_channels_per_group32_height32_width32_groups8_channel_lastFalse | 326.948 | 316.455 | 1.03
channel_shuffle_batch_size4_channels_per_group32_height32_width64_groups4_channel_lastTrue | 1423.315 | 317.068 | 4.49
channel_shuffle_batch_size4_channels_per_group32_height32_width64_groups4_channel_lastFalse | 325.531 | 316.596 | 1.03
channel_shuffle_batch_size4_channels_per_group32_height32_width64_groups8_channel_lastTrue | 6571.123 | 678.229 | 9.69
channel_shuffle_batch_size4_channels_per_group32_height32_width64_groups8_channel_lastFalse | 878.245 | 646.271 | 1.36
channel_shuffle_batch_size4_channels_per_group32_height64_width32_groups4_channel_lastTrue | 1423.888 | 316.824 | 4.49
channel_shuffle_batch_size4_channels_per_group32_height64_width32_groups4_channel_lastFalse | 325.770 | 316.317 | 1.03
channel_shuffle_batch_size4_channels_per_group32_height64_width32_groups8_channel_lastTrue | 6569.904 | 673.562 | 9.75
channel_shuffle_batch_size4_channels_per_group32_height64_width32_groups8_channel_lastFalse | 879.080 | 645.087 | 1.36
channel_shuffle_batch_size4_channels_per_group32_height64_width64_groups4_channel_lastTrue | 6501.307 | 673.263 | 9.66
channel_shuffle_batch_size4_channels_per_group32_height64_width64_groups4_channel_lastFalse | 862.646 | 645.630 | 1.34
channel_shuffle_batch_size4_channels_per_group32_height64_width64_groups8_channel_lastTrue | 13488.342 | 2107.488 | 6.40
channel_shuffle_batch_size4_channels_per_group32_height64_width64_groups8_channel_lastFalse | 2019.342 | 1911.166 | 1.06
channel_shuffle_batch_size4_channels_per_group64_height32_width32_groups4_channel_lastTrue | 1309.758 | 317.439 | 4.13
channel_shuffle_batch_size4_channels_per_group64_height32_width32_groups4_channel_lastFalse | 328.172 | 316.336 | 1.04
channel_shuffle_batch_size4_channels_per_group64_height32_width32_groups8_channel_lastTrue | 6578.720 | 660.705 | 9.96
channel_shuffle_batch_size4_channels_per_group64_height32_width32_groups8_channel_lastFalse | 902.144 | 645.365 | 1.40
channel_shuffle_batch_size4_channels_per_group64_height32_width64_groups4_channel_lastTrue | 6504.782 | 675.302 | 9.63
channel_shuffle_batch_size4_channels_per_group64_height32_width64_groups4_channel_lastFalse | 875.443 | 645.004 | 1.36
channel_shuffle_batch_size4_channels_per_group64_height32_width64_groups8_channel_lastTrue | 13474.683 | 2154.331 | 6.25
channel_shuffle_batch_size4_channels_per_group64_height32_width64_groups8_channel_lastFalse | 2066.495 | 1905.396 | 1.08
channel_shuffle_batch_size4_channels_per_group64_height64_width32_groups4_channel_lastTrue | 6479.559 | 736.724 | 8.80
channel_shuffle_batch_size4_channels_per_group64_height64_width32_groups4_channel_lastFalse | 878.409 | 649.855 | 1.35
channel_shuffle_batch_size4_channels_per_group64_height64_width32_groups8_channel_lastTrue | 13459.314 | 2053.354 | 6.55
channel_shuffle_batch_size4_channels_per_group64_height64_width32_groups8_channel_lastFalse | 2060.625 | 1906.264 | 1.08
channel_shuffle_batch_size4_channels_per_group64_height64_width64_groups4_channel_lastTrue | 13346.251 | 1993.075 | 6.70
channel_shuffle_batch_size4_channels_per_group64_height64_width64_groups4_channel_lastFalse | 2055.785 | 1935.230 | 1.06
channel_shuffle_batch_size4_channels_per_group64_height64_width64_groups8_channel_lastTrue | 30438.491 | 5696.887 | 5.34
channel_shuffle_batch_size4_channels_per_group64_height64_width64_groups8_channel_lastFalse | 4797.431 | 4646.504 | 1.03
channel_shuffle_batch_size8_channels_per_group32_height32_width32_groups4_channel_lastTrue | 850.684 | 317.521 | 2.68
channel_shuffle_batch_size8_channels_per_group32_height32_width32_groups4_channel_lastFalse | 328.374 | 316.374 | 1.04
channel_shuffle_batch_size8_channels_per_group32_height32_width32_groups8_channel_lastTrue | 3125.623 | 682.702 | 4.58
channel_shuffle_batch_size8_channels_per_group32_height32_width32_groups8_channel_lastFalse | 903.669 | 646.157 | 1.40
channel_shuffle_batch_size8_channels_per_group32_height32_width64_groups4_channel_lastTrue | 2950.647 | 672.118 | 4.39
channel_shuffle_batch_size8_channels_per_group32_height32_width64_groups4_channel_lastFalse | 879.654 | 645.399 | 1.36
channel_shuffle_batch_size8_channels_per_group32_height32_width64_groups8_channel_lastTrue | 13647.319 | 2020.811 | 6.75
channel_shuffle_batch_size8_channels_per_group32_height32_width64_groups8_channel_lastFalse | 2090.614 | 1932.686 | 1.08
channel_shuffle_batch_size8_channels_per_group32_height64_width32_groups4_channel_lastTrue | 2875.812 | 739.154 | 3.89
channel_shuffle_batch_size8_channels_per_group32_height64_width32_groups4_channel_lastFalse | 878.228 | 648.036 | 1.36
channel_shuffle_batch_size8_channels_per_group32_height64_width32_groups8_channel_lastTrue | 13564.547 | 2026.004 | 6.70
channel_shuffle_batch_size8_channels_per_group32_height64_width32_groups8_channel_lastFalse | 2093.294 | 1947.590 | 1.07
channel_shuffle_batch_size8_channels_per_group32_height64_width64_groups4_channel_lastTrue | 13354.635 | 2034.973 | 6.56
channel_shuffle_batch_size8_channels_per_group32_height64_width64_groups4_channel_lastFalse | 2043.774 | 1933.158 | 1.06
channel_shuffle_batch_size8_channels_per_group32_height64_width64_groups8_channel_lastTrue | 30528.643 | 5488.128 | 5.56
channel_shuffle_batch_size8_channels_per_group32_height64_width64_groups8_channel_lastFalse | 4801.484 | 4640.786 | 1.03
channel_shuffle_batch_size8_channels_per_group64_height32_width32_groups4_channel_lastTrue | 2727.968 | 644.440 | 4.23
channel_shuffle_batch_size8_channels_per_group64_height32_width32_groups4_channel_lastFalse | 894.660 | 645.652 | 1.39
channel_shuffle_batch_size8_channels_per_group64_height32_width32_groups8_channel_lastTrue | 13535.079 | 2131.735 | 6.35
channel_shuffle_batch_size8_channels_per_group64_height32_width32_groups8_channel_lastFalse | 2123.859 | 1937.743 | 1.10
channel_shuffle_batch_size8_channels_per_group64_height32_width64_groups4_channel_lastTrue | 13382.192 | 1997.564 | 6.70
channel_shuffle_batch_size8_channels_per_group64_height32_width64_groups4_channel_lastFalse | 2086.732 | 1920.518 | 1.09
channel_shuffle_batch_size8_channels_per_group64_height32_width64_groups8_channel_lastTrue | 30911.592 | 5818.629 | 5.31
channel_shuffle_batch_size8_channels_per_group64_height32_width64_groups8_channel_lastFalse | 4819.277 | 4721.406 | 1.02
channel_shuffle_batch_size8_channels_per_group64_height64_width32_groups4_channel_lastTrue | 13413.047 | 1930.201 | 6.95
channel_shuffle_batch_size8_channels_per_group64_height64_width32_groups4_channel_lastFalse | 2083.692 | 1940.887 | 1.07
channel_shuffle_batch_size8_channels_per_group64_height64_width32_groups8_channel_lastTrue | 30626.032 | 5529.039 | 5.54
channel_shuffle_batch_size8_channels_per_group64_height64_width32_groups8_channel_lastFalse | 4829.248 | 4657.726 | 1.04
channel_shuffle_batch_size8_channels_per_group64_height64_width64_groups4_channel_lastTrue | 30409.803 | 4680.921 | 6.50
channel_shuffle_batch_size8_channels_per_group64_height64_width64_groups4_channel_lastFalse | 4965.459 | 4666.142 | 1.06
channel_shuffle_batch_size8_channels_per_group64_height64_width64_groups8_channel_lastTrue | 65786.889 | 11573.771 | 5.68
channel_shuffle_batch_size8_channels_per_group64_height64_width64_groups8_channel_lastFalse | 10538.863 | 10004.425 | 1.05

mingfeima committed:
Update on "add channels last support for ChannelShuffle"




[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"




[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"




[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007052).

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007052).

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007052).

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007052).

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007052).

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima commented:
rebased.

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima commented:
rebased.

mingfeima commented:
TODO:
- [ ] rewrite channels last impl with fbgemm transpose.

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima commented:
This stack has been rebased, please help review :)

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007052).

VitalyFedyunin reviewed (commented):
Code looks good, please include tests

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

VitalyFedyunin commented:
Please rebase stack, landing tool requires it.

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

mingfeima commented:
updated the original test case from test_nn.py:

```bash
python test_nn.py TestNN.test_channel_shuffle
```

mingfeima commented:
The stack is newly rebased!

VitalyFedyunin commented:
Sorry, can you please rebase again, I'm getting ghstack errors `Internal error: couldn't understand base diff`

mingfeima committed:
Update on "add channels last support for ChannelShuffle"


Differential Revision: [D26007052](https://our.internmc.facebook.com/intern/diff/D26007052)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007052).

kimishpatel commented:
@mingfeima why does this PR introduce `native_channel_shuffle` in native_functions.yaml when it doesnt really need to provide aten op by the same name and thus be usable from torch namespace?

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3394,6 +3394,11 @@
     CPU: channel_shuffle
     QuantizedCPU: channel_shuffle_quantized_cpu
 
+- func: native_channel_shuffle(Tensor self, int groups) -> Tensor
```
Hey @VitalyFedyunin, it looks like this op was added without any python docs. For the upcoming release, can either mark the op as private or add docs for it, and cherry-pick into the release branch?

VitalyFedyunin commented:
@mingfeima could you please submit PR with python docstrings for this function.
